### ISSUE TITLE
Support for Matic Mainnet (Polygon)

### ISSUE THREAD

**Issue by MysticRyuujin** [emeraldpay/dshackle#86]
Can you add support for ChainID 137 for Polygon/Matic Mainnet?

They basically just run a geth clone called `bor` so it's all the same methods, but chainID needs to be 137...

**Comment by splix** [comment]
You're reading my mind :) Trying to setup Matic this weekend (saw your medium post on this topic btw, very helpful), and when I'll figure out how it all works I want to add support to Dshackle

**Comment by MysticRyuujin** [comment]
There MIGHT be more to it than my pull request, including tests lol

If I knew how to create a docker image from this I would...but you don't have a `Dockerfile` in the repo?

**Comment by splix** [comment]
Thank you.

Dockerfile is created during the build, but you can do it too. Try to run `./gradlew jibDockerBuild` command. It should download and install all the required stuff, compile the project and then make a local Docker build.

**Comment by MysticRyuujin** [comment]
Excellent, I will give it a shot. While, I have your attention, random question: I can share a redis cache between chains right? If I run with say, `ethereum` and `polygon` the redis cache won't get mixed up right?

**Comment by splix** [comment]
Yes, you can use shared Redis instance

**Comment by MysticRyuujin** [comment]
Hmm README instructions say to install JDK 11 but error is
`Could not target platform: 'Java SE 13' using tool chain: 'JDK 11 (11)'.` So I guess I need JDK13 now?

**Comment by splix** [comment]
Ah, yeah, changed to Java 13 as minimum supported recently, but I guess forgot to update the readme

**Comment by MysticRyuujin** [comment]
Pushed :smile:

**Comment by MysticRyuujin** [comment]
Ok I don't know how to solve the "Unresolved reference: POLYGON" messages haha

I seemed to build a docker image off the latest master with my changes, but it crashed on me, so I tried to switch to 0.9.6 release, and reapply my changes but I get those same errors as the CI.

I guess I'll just wait for you lol sorry I wish I could be more useful

**Comment by MysticRyuujin** [comment]
I have 2 endpoints up and running now, if you want/need to test against some nodes directly. I assume those public endpoints suffice, but just in case they don't, I can easily get you access to mine.

**Comment by splix** [comment]
Thank you @MysticRyuujin. My chain is almost synced, so I hope I'll start using it. I need to play with the network and software anyway, to better understand capabilities and other stuff. 

Though at the current point, I'm little bit disappointed, with the super slow sync, resource usage, non-existing docs, s3-based trusted copy, etc. Maybe it's just a bad first impression. Looking forward to actually try it.

**Comment by MysticRyuujin** [comment]
Yeah, make sure you take a look at my blog post again though, I updated it just the other day. When I originally posted it, I wasn't aware of the REST API server or the RabbitMQ server. I've updated it to include those. Cheers!

**Comment by MysticRyuujin** [comment]
Anything I can do to help move this forward?

**Comment by AJolly** [comment]
Would love to get this moving forward, getting reliable matic nodes gets complicated. Even running my own local nodes, I see cases where under heavy load they start falling behind. Same with even the paid providers out there. Would love to get Dshackle  up and running, without me having to compile from source.

**Comment by yorickdowne** [comment]
I’ve improved “own local node” with noatime, swap off, mitigations=off, and dedicated hardware with NVMe DC SSDs.

And I’d love a failover / balancing proxy that’s smart about sync state.

**Comment by splix** [comment]
Hi! Thank you for the request, and sorry for the delay, had to switch to other things. But now returned to the Dshackle and this feature is high priority right now.

**Comment by splix** [comment]
I gave Matic another try, but no luck in setting it up. It's the second time I'm trying to set up Matic, previous one was a little bit more successful, and I've got a semi-working-something, but it died later. 

Without a way to set up Matic on _my_ machine, I cannot do any development for it, so unfortunately I have to postpone this feature until the Matic is more production-ready.

**Comment by yorickdowne** [comment]
Hey I have a repo that dockerizes Matic and makes it pretty easy to spin up. Are you on discord or something? Happy to walk you through it and get Matic running for you locally on your machine. Once I’m off my phone let me also edit this and link the repo and give brief instructions.

**Comment by splix** [comment]
I actually tried to setup without Docker, just on a bare server. I use https://chasewright.com/how-to-run-a-polygon-matic-mainnet-node/ as a guide (and I highly recommend it), but still no luck

**Comment by MysticRyuujin** [comment]
I'm not sure I know what "production ready" means? They're in production. Is it kinda hard to setup, yes, but it's also working for hundreds of people? I'm also not sure I understand why you need a local node to test with? bor is a geth clone, there are no special methods to deal with? You can point to a public node to check response data? I can even give you access to my 2 private endpoints if it would help.

**Comment by yorickdowne** [comment]
If you're willing to try docker, this will start a Bor:

`git clone https://github.com/yorickdowne/eth-docker polygon-main && cd polygon-main && git fetch origin clone-chains && git checkout clone-chains && cp polygon-mainnet.env .env`

Optionally edit .env and set `HEIMDALL_ETH_RPC_URL` to the URL of an Ethereum node - this is not required however.

Then `./ethd start`

This grabs snapshots and starts the whole stack. Observe with `./ethd logs -f bor` and `./ethd logs -f heimdalld`

`./ethd` is just a wrapper around docker-compose. You can use docker-compose commands directly if you prefer.

Prerequesites for this are git, docker(.io or -ce) and docker-compose

The Bor that this starts has ports mapped to host for http and ws. Alternatively you can change `COMPOSE_FILE` to hook into a traefik, with `ext-network.yml`, or run a local traefik, with `traefik-cf.yml` or `traefik-aws.yml`. Holler if you want traefik and I'll link instructions for how that works with this tooling.

**Comment by yorickdowne** [comment]
> Is it kinda hard to setup, yes

Hence that whole docker idea. I needed a couple of these and wanted to do the work only once :)

**Comment by splix** [comment]
@MysticRyuujin 

Well, the fact that they don't even provide binaries yet says that the developers are not sure about its readiness for production use. Also, those hacks with the genesis files, misleading errors, and logs, etc., it shows that it's not mature. (And all of it is fine, nothing built overnight, just saying it have a long way before it's fully usable)

I need it locally because I want to see how it works, how it crashes, etc. I need to know what to answer people who email me and ask why Dshackle doesn't work on their side, how to debug their problem, and so on. I mean, if I say it works officially with Dshackle, I need to be sure it actually works and what to do when it doesn't. 

(NOTE: you can use Dshackle as is, just say it's an "ethereum" node in the upstreams config, it should work, I guess, but that's unofficial, i.e. is all up to you)

**Comment by MysticRyuujin** [comment]
What you suggest does NOT work because dshackle has static configurations for things like chainId, which can't be overwritten as far as I can tell. It would prevent us from mixing eth and matic in the same instance as well.

**Comment by yorickdowne** [comment]
@splix Let me know how you're making out with the docker repo, please. I am happy to assist with anything that might go bump.

**Comment by splix** [comment]
ah, right, chainId command would respond with a wrong value.

**Comment by splix** [comment]
@yorickdowne trying to find out how you made the configs. Can you point to it?

**Comment by yorickdowne** [comment]
Bor.yml has the details of how it gets started, and the environment variables are in polygon-mainnet.env. Was that the question?

**Comment by splix** [comment]
ok, well, I have problems with both bor and heimdall. I tried to reinstall each few times, each time different way, but never worked. So let's star with the heimdall, right now it says 

> E[2021-07-30|01:01:03.784] Error dialing seed                           module=p2p err="incompatible: Peer is on a different network. Got heimdall-137, expected heimdall-2BmrO1" seed=4fb1bc820088764a564d4f66bba1963d47d82329@44.232.55.71:26656 

and it seems starts to generate some invalid blocks like:

> I[2021-07-30|01:01:08.234] Executed block                               module=state height=145 validTxs=0 invalidTxs=0
> I[2021-07-30|01:01:08.237] Committed state                              module=state height=145 txs=0 appHash=9032B9383330807F7231C6DE3B168716EE4D34184CEE1FED29E3EC6401346958

I believe that these blocks are invalid because status gives:

> "latest_block_height": "142",
> "latest_block_time": "2021-07-29T23:00:48.203544347Z",
> "catching_up": false

which to my understanding is impossible.

Any ideas what could be wrong with the current setup?

**Comment by yorickdowne** [comment]
Hmm. That suggests your `HEIMDALL_CHAIN_ID=heimdall-137` might not be set. All those variables are in `polygon-mainnet.env`. As well, when you look at `bor.yml` and the heimdalld service, you'll see it does a heimdalld --init with the chainid, grabs a snapshot from https://snapshots.matic.today/, and it grabs a genesis file from the polygon github.

There's a lot going on, which is why it's all encapsulated in docker.

**Comment by splix** [comment]
When I tried to run with the snapshot it was failing with:

> [2021-07-29|21:06:24.091] Unable to get checkpoint                     module=checkpoint error="No checkpoint found in buffer"

So I decided to start from scratch. If you know how to fix that error, that would be even better

**Comment by splix** [comment]
Can you point to exact Dockerfile?

**Comment by yorickdowne** [comment]
I am using a docker-compose file, `bor.yml`. It uses published images, `maticnetwork/bor` and `maticnetwork/heimdall`, plus a rabbitmq. Hence no Dockerfile as such. I used https://chasewright.com/how-to-run-a-polygon-matic-mainnet-node/ and an official docker-compose file that I have a bear of a time finding again.

**Comment by splix** [comment]
It seems I'm looking into a wrong repo. There is no `bor.xml` or anything that looks like a matic specific config at https://github.com/yorickdowne/eth-docker

**Comment by yorickdowne** [comment]
Look at the `clone-chains` branch please in that repo. eth-docker is meant for Ethereum stakers, I didn't want to overload it with various stuff, so all the clone things are in that branch - Polygon, HECO, Fantom.

**Comment by splix** [comment]
ah, I see, thank you

**Comment by yorickdowne** [comment]
I spun up a fresh copy overnight just to make sure these snapshots were good. Looks fine; heimdalld is catching up and struggling with peers as it usually does; bor is getting data from heimdalld and slowly moving forward from the snapshot date, the 22nd. Give it a day and they should both be synced.
Edit: I am seeing the occasional panic in heimdalld re "mainchain RPC fails and then my own RPC fails", so just to make it happier, I gave it a mainchain endpoint to talk to, then restarted it. It's behaving much better now. Infura should work for that.

**Comment by splix** [comment]
@yorickdowne thank you, it seems I made it working. Heimdall is fully synced now, waiting for Bor, but all looks good so far

**Comment by MysticRyuujin** [comment]
So I got a little impatient for a release and I built the latest off master and it seems like it's mostly working but I did get a weird error:

```
2021-20-08 18:30:37.924 | WARN  | WsFactory$EthereumWs | Connection dropped to wss://rpc-mainnet.maticvigil.com/ws/v1/<redacted>. Error: Invalid UTF-8 middle byte 0x2b
 at [Source: (io.netty.buffer.ByteBufInputStream); line: 1, column: 4]
2021-20-08 18:30:38.835 | WARN  | WsFactory$EthereumWs | Connection dropped to wss://rpc-mainnet.maticvigil.com/ws/v1/<redacted>. Error: Illegal character ((CTRL-CHAR, code 16)): only regular white space (\r, \n, \t) is allowed between tokens
 at [Source: (io.netty.buffer.ByteBufInputStream); line: 1, column: 2
 ```
 
 This could be something weird that provider is doing, idk, I need to add my own bor nodes to check but I thought I'd mention it.

**Comment by czarly** [comment]
guys let me say thank you for your efforts. I gave up on docker for polygon a while back and just used their official ansible scripts but I like the docker approach much better and may just switch to what you came up with. I'm also looking forward to dshackle support.

**Comment by yorickdowne** [comment]
Heads-up that Bor just received the hotfix for EVM vuln from upstream Geth. Time to update both Geth and Bor.

**Comment by cventastic** [comment]
Hello.
I'm new to Blockchain technology and i can't figure out why bor can't get in sync. 
Everything worked fine until bor finished with extracting the snapshot. Now it tries to sync from Peers and i get:
`
bor_1        | Error: invalid merkle root (remote: 744fb39148037c50a769e4eb2810fcc7aaa75855bc183cbfbf1cfcf2b59d7e01 local: 90fc6d1a054bc097a8f24fd9293bab52160217e84bf9480aef654c4e60d1198b)
`
and i don't know what could be the reason for this. Any help would be appreciated.

**Comment by splix** [comment]
Available with the latest release v0.10
